tests-java8/src/test/java/com/dslplatform/json/ComplexGenericsTest.java

```diff
@@ -0,0 +1,222 @@
+        }
+
+        @Override
+        public String getType() {
+            return "first";
+        }
+    }
+
+    @CompiledJson(deserializeName = "second")
+    public static class SecondChild2 extends Parent2<String> {
+
+        private Integer intValue;
+
+        public Integer getIntValue() {
+            return intValue;
+        }
+
+        public void setIntValue(Integer intValue) {
+            this.intValue = intValue;
+        }
+
+        @Override
+        public String getType() {
+            return "second";
+        }
+    }
+
+    @CompiledJson
+    public static abstract class Parent3<T> {
+
+        private T prop;
+
+        public T getProp() {
+            return prop;
+        }
+
+        public void setProp(T prop) {
+            this.prop = prop;
+        }
+    }
+
+    @CompiledJson
+    public static class FirstChild3 extends Parent3<Long> {
+
+        private Boolean boolValue;
+
+        public Boolean getBoolValue() {
+            return boolValue;
+        }
+
+        public void setBoolValue(Boolean boolValue) {
+            this.boolValue = boolValue;
+        }
+    }
+
+    @CompiledJson
+    public static class SecondChild3 extends Parent3<String> {
+
+        private Integer intValue;
+
+        public Integer getIntValue() {
+            return intValue;
+        }
+
+        public void setIntValue(Integer intValue) {
+            this.intValue = intValue;
+        }
+    }
+
+    private final DslJson<Object> dslJson = new DslJson<>();
+
+    @Test
+    public void inheritanceWithGenericsTestAndCustomDiscriminator() throws IOException {
+
+        byte[] bytes = "{\"type\":\"first\",\"prop\":1,\"boolValue\":true}".getBytes();
+        Parent parent = dslJson.deserialize(Parent.class, bytes, bytes.length);
+
+        assertNotNull(parent);
+        assertEquals("first", parent.getType());
+        assertTrue(parent instanceof FirstChild);
+        FirstChild child = (FirstChild) parent;
+        assertEquals(1L, child.getProp().longValue());
+        assertTrue(child.getBoolValue());
+
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        dslJson.serialize(child, outputStream);
+
+        assertEquals("{\"type\":\"first\",\"boolValue\":true,\"prop\":1}", outputStream.toString());
+    }
+
+    @Test
+    public void inheritanceWithGenericsTest() throws IOException {
+
+        byte[] bytes = "{\"$type\":\"first\",\"prop\":1,\"boolValue\":true}".getBytes();
+        Parent2 parent = dslJson.deserialize(Parent2.class, bytes, bytes.length);
+
+        assertNotNull(parent);
+        assertEquals("first", parent.getType());
+        assertTrue(parent instanceof FirstChild2);
+        FirstChild2 child = (FirstChild2) parent;
+        assertEquals(1L, child.getProp().longValue());
+        assertTrue(child.getBoolValue());
+
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        dslJson.serialize(child, outputStream);
```

Comment:
The previous test where you explicitly specified the type was the expected behavior.
This doesn't look like the expected behavior to me unless you explicitly add that discriminator also to the class annotation

Reply:
Agree.